File: src/app/appearance.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { useCallback, useEffect, useRef, useState, type CSSProperties } from "react";
import { debugLog, isTauriRuntime, safeInvoke } from "./platform";

/**
 * The class the ghost stylesheet keys off.
 *
 * `applyAppearance` is the only place allowed to put it on an element, and
 * `<html>` is the only element it goes on. A component adding a second copy
 * driven by its own state is how the island ended up ignoring every slider the
 * settings panel wrote — the two copies read different sources and disagreed.
 */
export const GHOST_ROOT_CLASS = "meetly-ghost";

/**
 * Single source of truth for how the overlay looks and how big it is.
 *
 * These settings live on disk (Rust side) instead of in `localStorage`:
 * every Tauri window is its own webview, and the island never received the
 * `storage` events the settings window fired — which is why tuning used to
 * show up in the preview only. Rust persists the value and broadcasts
 * `appearance_changed`, so all windows update the moment a slider moves.
 */

export type AppearanceSettings = {
  ghostEnabled: boolean;
  /** Grey level of ghost text, 0–255, rendered as an `R G B` triplet. */
  ink: number;
  strong: number;
  base: number;
  soft: number;
  /** Scales the whole UI and the windows that host it. */
  uiScale: number;
  /** Collapsed island width in CSS pixels, before `uiScale`. */
  islandWidth: number;
  panelWidth: number;
  panelHeight: number;
};

export const APPEARANCE_EVENT = "appearance_changed";
export const APPEARANCE_STORAGE_KEY = "meetly.appearance";

export const DEFAULT_APPEARANCE: AppearanceSettings = {
  ghostEnabled: true,
  ink: 130,
  strong: 0.82,
  base: 0.66,
  soft: 0.46,
  uiScale: 1,
  islandWidth: 600,
  panelWidth: 920,
  panelHeight: 600,
};

export const APPEARANCE_LIMITS = {
  ink: { min: 40, max: 235, step: 1 },
  opacity: { min: 0.05, max: 1, step: 0.01 },
  uiScale: { min: 0.7, max: 1.6, step: 0.05 },
  islandWidth: { min: 320, max: 1400, step: 10 },
  panelWidth: { min: 560, max: 1600, step: 10 },
  panelHeight: { min: 320, max: 1100, step: 10 },
} as const;

function clamp(value: number, min: number, max: number): number {
  if (!Number.isFinite(value)) return min;
  return Math.min(max, Math.max(min, value));
}

/** Mirrors the clamping Rust applies, so previews never lie about the result. */
export function sanitizeAppearance(value: Partial<AppearanceSettings> | null): AppearanceSettings {
  const source = value ?? {};
  const number = (raw: unknown, range: { min: number; max: number }, fallback: number) =>
    clamp(typeof raw === "number" ? raw : fallback, range.min, range.max);

  return {
    ghostEnabled: source.ghostEnabled !== false,
    ink: number(source.ink, APPEARANCE_LIMITS.ink, DEFAULT_APPEARANCE.ink),
    strong: number(source.strong, APPEARANCE_LIMITS.opacity, DEFAULT_APPEARANCE.strong),
    base: number(source.base, APPEARANCE_LIMITS.opacity, DEFAULT_APPEARANCE.base),
    soft: number(source.soft, APPEARANCE_LIMITS.opacity, DEFAULT_APPEARANCE.soft),
    uiScale: number(source.uiScale, APPEARANCE_LIMITS.uiScale, DEFAULT_APPEARANCE.uiScale),
    islandWidth: number(
      source.islandWidth,
      APPEARANCE_LIMITS.islandWidth,
      DEFAULT_APPEARANCE.islandWidth,
    ),
    panelWidth: number(
      source.panelWidth,
      APPEARANCE_LIMITS.panelWidth,
      DEFAULT_APPEARANCE.panelWidth,
    ),
    panelHeight: number(
      source.panelHeight,
      APPEARANCE_LIMITS.panelHeight,
      DEFAULT_APPEARANCE.panelHeight,
    ),
  };
}

function readLocal(): AppearanceSettings | null {
  try {
    const raw = window.localStorage.getItem(APPEARANCE_STORAGE_KEY);
    return raw ? sanitizeAppearance(JSON.parse(raw) as Partial<AppearanceSettings>) : null;
  } catch {
    return null;
  }
}

function writeLocal(settings: AppearanceSettings): void {
  try {
    window.localStorage.setItem(APPEARANCE_STORAGE_KEY, JSON.stringify(settings));
  } catch {
    /* Storage is best-effort; the settings still apply for the current session. */
  }
}

/**
 * Last known value, shared by every hook in this window. `patchAppearance`
 * needs a full payload to send, and reading it from here avoids an extra
 * round trip on every slider tick.
 */
let cached: AppearanceSettings | null = null;

export async function loadAppearance(): Promise<AppearanceSettings> {
  if (isTauriRuntime()) {
    try {
      const settings = await safeInvoke<AppearanceSettings>("get_appearance");
      if (settings) {
        cached = sanitizeAppearance(settings);
        writeLocal(cached);
        return cached;
      }
    } catch {
      /* Fall through to the local copy when the command is unavailable. */
    }
  }

  cached = readLocal() ?? sanitizeAppearance(null);
  return cached;
}

export async function saveAppearance(
  settings: AppearanceSettings,
): Promise<AppearanceSettings> {
  const next = sanitizeAppearance(settings);
  cached = next;
  writeLocal(next);

  if (!isTauriRuntime()) {
    return next;
  }

  try {
    const saved = await safeInvoke<AppearanceSettings>("save_appearance", { settings: next });
    if (saved) {
      cached = sanitizeAppearance(saved);
    }
  } catch (error) {
    console.error("Failed to save appearance:", error);
  }

  return cached;
}

/** Applies a partial change on top of the last known settings. */
export async function patchAppearance(
  patch: Partial<AppearanceSettings>,
): Promise<AppearanceSettings> {
  const base = cached ?? (await loadAppearance());
  return saveAppearance({ ...base, ...patch });
}

/**
 * Publishes the settings onto the document: the ghost class, the four ghost
 * CSS variables and the UI zoom.
 *
 * Zoom goes on `<html>` rather than a wrapper because the layout is built
 * from `100vh`/`100vw`; CSS zoom rescales the initial containing block, so
 * the panel keeps filling the window exactly at any scale.
 */
export function applyAppearance(
  settings: AppearanceSettings,
  options?: { zoom?: boolean },
): void {
  const root = document.documentElement;
  if (!root) return;

  root.classList.toggle(GHOST_ROOT_CLASS, settings.ghostEnabled);

  const ink = Math.round(settings.ink);
  root.style.setProperty("--ghost-ink", `${ink} ${ink} ${ink}`);
  root.style.setProperty("--ghost-strong", settings.strong.toFixed(2));
  root.style.setProperty("--ghost-base", settings.base.toFixed(2));
  root.style.setProperty("--ghost-soft", settings.soft.toFixed(2));

  void reportRender(windowLabel(), settings);

  // Inside Tauri the Rust side owns zoom (`WKWebView.pageZoom`, applied to the
  // window and its layout together). CSS zoom is only the fallback for a plain
  // browser session, where no native window exists.
  if (!isTauriRuntime() && options?.zoom !== false) {
    root.style.zoom = Math.abs(settings.uiScale - 1) < 0.001 ? "" : String(settings.uiScale);
  }
}

/** Which window this webview is: the log is the only way to see the overlays. */
function windowLabel(): string {
  const path = window.location.pathname;
  if (path.includes("voice-overlay")) return "voice";
  if (path.includes("settings")) return "settings";
  return "island";
}

/**
 * Stand-in for real body copy.
 *
 * The live panels only exist once a session is open, and the settings preview
 * is deliberately exempt from the ghost rules, so while the settings surface is
 * up neither one can say what an actual transcript line resolves to — which is
 * exactly the case that hid the original bug. This probes the cascade itself:
 * same classes, same ancestor, real computed colour, whatever the view is
 * showing. Appended, read and removed inside one tick, so it never paints.
 */
function probeColor(outerClass: string, tag: string): string {
  const holder = document.createElement("div");
  holder.className = outerClass;
  const inner = document.createElement(tag);
  inner.textContent = "probe";
  holder.appendChild(inner);

  // Directly under <html>: it inherits the ghost class like a real panel, and
  // it stays outside .workspace-settings, which restores its own light colours.
  document.documentElement.appendChild(holder);
  try {
    return getComputedStyle(inner).color;
  } finally {
    holder.remove();
  }
}

/**
 * The overlay windows are content-protected, so no screenshot can show whether
 * a slider actually changed anything. Writing the values the browser reports
 * back — not the ones we sent — is the only objective evidence.
 */
function reportRender(label: string, settings: AppearanceSettings): void {
  if (!isTauriRuntime()) return;

  const root = document.documentElement;
  const styles = getComputedStyle(root);
  const read = (name: string) => styles.getPropertyValue(name).trim() || "?";

  const sample = (selector: string, { insidePreview }: { insidePreview: boolean }) => {
    const found = Array.from(document.querySelectorAll(selector)).find((element) =>
      insidePreview
        ? Boolean(element.closest(".ghost-preview-stage"))
        : !element.closest(".ghost-preview-stage"),
    );
    return found ? getComputedStyle(found).color : "no-target";
  };

  // The live panels and the settings preview are reported separately: the
  // preview is the only thing visible while the settings surface is open, and
  // it wears the ghost class on purpose, so lumping it in with the real panels
  // once reported a healthy overlay no matter what the panels were doing.
  const live = sample(".agent-markdown, .transcript-line, .meetly-island", { insidePreview: false });
  const preview = sample(".agent-markdown", { insidePreview: true });
  const shell = document.querySelector(".workspace-settings");

  // `body` is the tier the agent answers render in, `meta` the timestamps and
  // hints. Together they prove the sliders reach the real text rules and not
  // just the preview, whatever is on screen.
  const body = probeColor("agent-markdown", "p");
  const meta = probeColor("agent-message-meta", "span");

  debugLog(
    `[render:${label}] ghost-class=${root.classList.contains(GHOST_ROOT_CLASS)} ` +
      `ink=${read("--ghost-ink")} base=${read("--ghost-base")} ` +
      `strong=${read("--ghost-strong")} soft=${read("--ghost-soft")} ` +
      `| asked ink=${Math.round(settings.ink)} ` +
      `| body=${body} meta=${meta} ` +
      `| live=${live} preview=${preview} ` +
      `settings-shell=${shell ? getComputedStyle(shell).backgroundColor : "no-shell"}`,
  );
}

/** Slider-friendly inline style for the ghost preview surfaces. */
export function ghostPreviewStyle(settings: AppearanceSettings): CSSProperties {
  const ink = Math.round(settings.ink);
  return {
    "--ghost-ink": `${ink} ${ink} ${ink}`,
    "--ghost-strong": settings.strong.toFixed(2),
    "--ghost-base": settings.base.toFixed(2),
    "--ghost-soft": settings.soft.toFixed(2),
  } as CSSProperties;
}

export type AppearanceStore = {
  settings: AppearanceSettings;
  isReady: boolean;
  update: (patch: Partial<AppearanceSettings>) => void;
};

/**
 * Reads the settings on mount, keeps them in sync with every other window via
 * the `appearance_changed` event, and throttles writes so dragging a slider
 * does not flood the IPC channel.
 */
export function useAppearance(): AppearanceStore {
  const [settings, setSettings] = useState<AppearanceSettings>(
    () => cached ?? readLocal() ?? DEFAULT_APPEARANCE,
  );
  const [isReady, setIsReady] = useState(false);
  const latest = useRef(settings);
  const pending = useRef<ReturnType<typeof setTimeout> | null>(null);
  const sent = useRef<string | null>(null);

  latest.current = settings;

  useEffect(() => {
    void loadAppearance().then((next) => {
      setSettings(next);
      setIsReady(true);
    });
  }, []);

  useEffect(() => {
    if (!isTauriRuntime()) return;

    let unlisten: UnlistenFn | null = null;
    void listen<AppearanceSettings>(APPEARANCE_EVENT, (event) => {
      const next = sanitizeAppearance(event.payload);
      debugLog(`[event:${windowLabel()}] got ghost=${next.ghostEnabled} ink=${Math.round(next.ink)}`);
      setSettings((current) =>
        JSON.stringify(current) === JSON.stringify(next) ? current : next,
      );
    }).then((fn) => {
      unlisten = fn;
    });

    return () => {
      unlisten?.();
    };
  }, []);

  const commit = useCallback((next: AppearanceSettings) => {
    cached = next;
    // First change lands immediately (snappy feedback), the rest are
    // coalesced so a drag produces a handful of saves instead of hundreds.
    if (pending.current) return;

    sent.current = JSON.stringify(next);
    void saveAppearance(next);

    pending.current = setTimeout(() => {
      pending.current = null;

      // The trailing write only exists to flush the tail of a drag. Re-sending
      // an unchanged payload would be pure waste: every save makes the Rust
      // side re-measure and resize the island, so a single slider tick used to
      // move the window twice.
      const snapshot = JSON.stringify(latest.current);
      if (snapshot === sent.current) return;

      sent.current = snapshot;
      void saveAppearance(latest.current);
    }, 120);
  }, []);

  const update = useCallback(
    (patch: Partial<AppearanceSettings>) => {
      // Derived from the ref, not inside a state updater: updaters must stay
      // pure, and React may run them twice.
      const next = sanitizeAppearance({ ...latest.current, ...patch });
      latest.current = next;
      setSettings(next);
      commit(next);
    },
    [commit],
  );

  useEffect(
    () => () => {
      if (pending.current) clearTimeout(pending.current);
    },
    [],
  );

  return { settings, isReady, update };
}

/**
 * `useAppearance` plus the DOM side effects an overlay window needs. The
 * settings window uses the plain hook: it previews values instead of wearing
 * them.
 */
export function useOverlayAppearance(): AppearanceStore {
  const store = useAppearance();

  useEffect(() => {
    applyAppearance(store.settings);
  }, [store.settings]);

  return store;
}

File: src-tauri/src/appearance.rs
//! Window appearance settings shared by every webview.
//!
//! Ghost mode started life as a localStorage flag, but each Tauri window is a
//! separate WKWebView and does not reliably share `localStorage` or deliver
//! `storage` events across windows — so a slider dragged in Settings never
//! reached the floating island. These settings live on disk and are broadcast
//! with a Tauri event, which every window receives.

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::{fs, path::PathBuf};
use tauri::{AppHandle, Emitter, Manager};

pub const APPEARANCE_EVENT: &str = "appearance_changed";

const INK_RANGE: (f64, f64) = (40.0, 235.0);
const OPACITY_RANGE: (f64, f64) = (0.05, 1.0);
const UI_SCALE_RANGE: (f64, f64) = (0.7, 1.6);
const ISLAND_WIDTH_RANGE: (f64, f64) = (320.0, 1400.0);
const PANEL_WIDTH_RANGE: (f64, f64) = (560.0, 1600.0);
const PANEL_HEIGHT_RANGE: (f64, f64) = (320.0, 1100.0);

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppearanceSettings {
    pub ghost_enabled: bool,
    /// Grey level of ghost text (0–255), rendered as an `R G B` triplet.
    pub ink: f64,
    pub strong: f64,
    pub base: f64,
    pub soft: f64,
    /// Scales the whole overlay UI and the windows that host it.
    pub ui_scale: f64,
    /// Collapsed island width, in CSS pixels before `ui_scale`.
    pub island_width: f64,
    pub panel_width: f64,
    pub panel_height: f64,
}

impl Default for AppearanceSettings {
    fn default() -> Self {
        Self {
            ghost_enabled: true,
            ink: 130.0,
            strong: 0.82,
            base: 0.66,
            soft: 0.46,
            ui_scale: 1.0,
            island_width: 600.0,
            panel_width: 920.0,
            panel_height: 600.0,
        }
    }
}

impl AppearanceSettings {
    /// Clamps every field so a corrupted file or a hostile IPC payload can
    /// never produce an invisible, zero-sized or off-screen window.
    pub fn sanitized(self) -> Self {
        Self {
            ghost_enabled: self.ghost_enabled,
            ink: clamp(self.ink, INK_RANGE),
            strong: clamp(self.strong, OPACITY_RANGE),
            base: clamp(self.base, OPACITY_RANGE),
            soft: clamp(self.soft, OPACITY_RANGE),
            ui_scale: clamp(self.ui_scale, UI_SCALE_RANGE),
            island_width: clamp(self.island_width, ISLAND_WIDTH_RANGE),
            panel_width: clamp(self.panel_width, PANEL_WIDTH_RANGE),
            panel_height: clamp(self.panel_height, PANEL_HEIGHT_RANGE),
        }
    }
}

fn clamp(value: f64, (min, max): (f64, f64)) -> f64 {
    if value.is_finite() {
        value.max(min).min(max)
    } else {
        min
    }
}

fn appearance_path() -> Result<PathBuf> {
    Ok(crate::app_state::state_dir()?.join("appearance.json"))
}

/// Reads the settings, falling back to defaults for missing or invalid files.
pub fn load() -> AppearanceSettings {
    let Ok(path) = appearance_path() else {
        return AppearanceSettings::default();
    };
    if !path.exists() {
        return AppearanceSettings::default();
    }

    let Ok(bytes) = fs::read(&path) else {
        return AppearanceSettings::default();
    };

    serde_json::from_slice::<AppearanceSettings>(&bytes)
        .map(|settings| settings.sanitized())
        .unwrap_or_default()
}

fn store(settings: &AppearanceSettings) -> Result<()> {
    let bytes =
        serde_json::to_vec_pretty(settings).context("Failed to serialize appearance settings")?;
    fs::write(appearance_path()?, bytes).context("Failed to write appearance settings")
}

#[tauri::command]
pub fn get_appearance() -> Result<AppearanceSettings, String> {
    Ok(load())
}

#[tauri::command]
pub fn save_appearance(
    app: AppHandle,
    settings: AppearanceSettings,
) -> Result<AppearanceSettings, String> {
    let next = settings.sanitized();
    store(&next).map_err(|error| error.to_string())?;

    // The overlay windows are content-protected, so there is no screenshot to
    // check when a setting "does nothing" — this line is the only evidence.
    let _ = crate::debug_log::append(&format!(
        "[appearance] saved ghost={} ink={} strong={:.2} base={:.2} soft={:.2} scale={:.2} island={} panel={}x{}",
        next.ghost_enabled,
        next.ink,
        next.strong,
        next.base,
        next.soft,
        next.ui_scale,
        next.island_width,
        next.panel_width,
        next.panel_height,
    ));

    // Resize right away so dragging a slider moves the real window, not just
    // the preview. Content protection and page zoom follow the settings for
    // every overlay window, not only the one that owns the toggle.
    crate::window::apply_island_metrics(&app);
    crate::window::apply_overlay_metrics(&app);
    apply_content_protection(&app, next.ghost_enabled);
    apply_zoom(&app, next.ui_scale);

    app.emit(APPEARANCE_EVENT, next)
        .map_err(|error| error.to_string())?;

    Ok(next)
}

fn apply_content_protection(app: &AppHandle, enabled: bool) {
    for label in OVERLAY_WINDOWS {
        if let Some(window) = app.get_webview_window(label) {
            let _ = window.set_content_protected(enabled);
        }
    }
}

/// The floating windows — the settings window keeps a fixed, readable size.
const OVERLAY_WINDOWS: [&str; 2] = ["island", "voice-overlay"];

/**
 * Scales the page itself (`WKWebView.pageZoom` on macOS), so text, spacing and
 * every CSS pixel grow together. The windows are resized by the same factor,
 * which keeps `100vh` layouts filling the window exactly.
 */
pub fn apply_zoom(app: &AppHandle, scale: f64) {
    for label in OVERLAY_WINDOWS {
        if let Some(window) = app.get_webview_window(label) {
            let _ = window.set_zoom(scale);
        }
    }
}

/// Applies everything a freshly created window needs.
pub fn apply_to_windows(app: &AppHandle) {
    let settings = load();
    apply_content_protection(app, settings.ghost_enabled);
    apply_zoom(app, settings.ui_scale);
}

#[cfg(test)]
mod tests {
    use super::{clamp, AppearanceSettings};

    #[test]
    fn out_of_range_values_are_clamped() {
        let settings = AppearanceSettings {
            ink: 900.0,
            strong: 5.0,
            base: -1.0,
            soft: f64::NAN,
            ui_scale: 0.0,
            island_width: 10.0,
            panel_width: 99999.0,
            panel_height: 0.0,
            ..AppearanceSettings::default()
        }
        .sanitized();

        assert_eq!(settings.ink, 235.0);
        assert_eq!(settings.strong, 1.0);
        assert_eq!(settings.base, 0.05);
        assert_eq!(settings.soft, 0.05);
        assert_eq!(settings.ui_scale, 0.7);
        assert_eq!(settings.island_width, 320.0);
        assert_eq!(settings.panel_width, 1600.0);
        assert_eq!(settings.panel_height, 320.0);
    }

    #[test]
    fn non_finite_values_fall_back_to_the_minimum() {
        assert_eq!(clamp(f64::INFINITY, (1.0, 2.0)), 1.0);
    }
}
